### PR TITLE
draft / RFC: Add support for application defined events

### PR DESCRIPTION
**Pull request by rcoh**
The main challenge is access to string interning which I solved by introducing `Encodable`. A side benefit is that `Encodeable` is dyn Safe which is convenient.

### CODE REVIEW

**Review by jlizen — APPROVED** [review]
This seems reasonable to me on first read-through. I say, go ahead and implement it, and we can iterate on first usage if it isn't quite right.

File: dial9-tokio-telemetry/tests/end_to_end.rs [hunk 1]
```diff
@@ -143,3 +143,56 @@ fn task_terminate_events_are_captured() {
         "expected at least {N} TaskTerminate events, got {terminate_count}"
     );
 }
+
```
Comment (jlizen) [inline dial9-tokio-telemetry/tests/end_to_end.rs:146]: We probably should have a test for the advanced Encodable path (manual impl with string interning) in the test suite.


Also no test for the edge case where encode is called multiple times within a single Encodable::encode impl

File: dial9-tokio-telemetry/src/telemetry/buffer.rs [hunk 2]
```diff
@@ -17,89 +17,129 @@ use std::sync::atomic::{AtomicU64, Ordering};
...
+/// In the example above, the trace will contain events named `"HttpRequestWire"`,
+/// not `"HttpRequest"`.
+pub trait Encodable {
+    /// Encode this event into the thread-local trace buffer.
+    ///
+    /// Implementations should call [`ThreadLocalEncoder::encode`] exactly once.
+    /// Each `encode` call is counted as one event for buffer flush decisions;
+    /// calling `encode` multiple times will produce multiple wire events but
+    /// only one event will be counted.
+    fn encode(&self, encoder: &mut ThreadLocalEncoder<'_>);
```
Comment (jlizen) [inline dial9-tokio-telemetry/src/telemetry/buffer.rs:122]: This means that `event_count` will no longer match what is sent on the wire, do we care?

File: dial9-tokio-telemetry/src/telemetry/buffer.rs [hunk 3]
```diff
@@ -17,89 +17,129 @@ use std::sync::atomic::{AtomicU64, Ordering};
...
+///             method,
+///             status: self.status,
+///         });
+///     }
+/// }
+/// ```
+///
+/// # Wire event naming
+///
+/// The event name in the trace comes from the struct passed to
```
Comment (jlizen) [inline dial9-tokio-telemetry/src/telemetry/buffer.rs:111]: This might be a bit surprising, would it be worth supporting custom name method? I'm fine to leave it as-is for now though and see if anybody asks.

File: dial9-tokio-telemetry/README.md [hunk 4]
```diff
@@ -107,6 +107,39 @@ runtime.block_on(async {
...
+        latency_us: 1500,
+    },
+    &handle,
+);
+# }
+```
+
+For events with repeated string values (HTTP methods, endpoint paths, etc.), implement `Encodable` manually to use string interning — see [`examples/custom_events.rs`](/dial9-tokio-telemetry/examples/custom_events.rs) for a complete example showing both patterns.
+
+Custom events are encoded into the same thread-local buffer as built-in events (~100–200 ns per call) and appear in the trace viewer alongside poll/park/wake events.
```
Comment (jlizen) [inline dial9-tokio-telemetry/README.md:141]: If we are going to claim this, should we have benches for the custom events?

**Review by jlizen — APPROVED** [review]
